File: UVA/Random/357 - Let Me Count The Ways.cpp
#include <assert.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <algorithm>
#include <numeric>
#include <functional>
#include <bitset>
#include <vector>
#include <list>
#include <set>
#include <map>
#include <queue>
#include <stack>
#include <string>
#include <sstream>
#include <iostream>
#include <limits.h>
using namespace std;

typedef long long int64;
#define SZ(a) ((int)a.size())
#define FOR(i, s, e) for (int i = s; i < e; ++i)
#define REP(i, e) for (int i = 0; i < e; ++i)
#define FOREACH(i,c) for (__typeof((c).begin()) i=(c).begin();i!=(c).end();i++)
#define ALL(c) c.begin(), c.end()
#define MAX 30005

int64 dp[MAX];
int64 coins[] = {1, 5, 10, 25, 50};

int main()
{
#ifndef ONLINE_JUDGE
	freopen("in.txt", "r", stdin);
#endif
	dp[0] = dp[1] = 1;
	REP(i, 5) for (int val = 1; val + coins[i] < MAX; val++)
		dp[val + coins[i]] += dp[val];
	int n;
	while (scanf("%d", &n) != EOF)
		if (dp[n + 1] == 1)
			printf("There is only 1 way to produce %d cents change.\n", n);
		else
			printf("There are %lld ways to produce %d cents change.\n", dp[n + 1], n);
	return 0;
}

File: UVA/Competitive Programming/Chapter 3/Dynamic Programming/Coin Change/357 - Let Me Count The Ways.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.Arrays;
import java.util.Scanner;

public class Main {
    static final int MAX = 30000 + 1;
    static final int[] coins = {1, 5, 10, 25, 50};

    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        long[] ways = new long[MAX];
        ways[0] = 1;
        for (int i = 0; i < coins.length; ++i)
            for (int j = coins[i]; j < MAX; ++j)
                if (j - coins[i] >= 0)
                    ways[j] += ways[j - coins[i]];
        while (cin.hasNext()) {
            int n = cin.nextInt();
            long res = ways[n];
            if (res == 1)
                System.out.printf("There is only 1 way to produce %d cents change.\n", n);
            else
                System.out.printf("There are %d ways to produce %d cents change.\n", res, n);
        }
    }

    private static void debug(Object...os) {
        System.out.println(Arrays.deepToString(os));
    }
}